### PR TITLE
Constify model

### PR DESCRIPTION
**Pull request by hendrik-skubch**

### CODE REVIEW

File: alica_engine/include/engine/PlanRepository.h [hunk 1]
```diff
@@ -42,59 +43,100 @@ class PlanRepository {
     PlanRepository();
```
Comment (gauthamm) [inline alica_engine/include/engine/PlanRepository.h]: https://msdn.microsoft.com/en-us/library/dn457344.aspx it seems mentioning constructor as default and removing the empty constructor definition makes it more efficient

File: alica_engine/include/engine/SimplePlanTree.h [hunk 2]
```diff
@@ -57,7 +57,7 @@ class SimplePlanTree {
      * The timestamp denoting when this tree was received.
      */
     long receiveTime;
-    list<long> stateIds;
+    std::list<long> stateIds;
```
Comment (gauthamm) [inline alica_engine/include/engine/SimplePlanTree.h]: int64_t

File: alica_engine/include/engine/PlanRepository.h [hunk 3]
```diff
@@ -42,59 +43,100 @@ class PlanRepository {
...
+    const Accessor<Task> getTasks() const { return Accessor<Task>(_tasks); }
+    const Accessor<Transition> getTransitions() const { return Accessor<Transition>(_transitions); }
+    const Accessor<Variable> getVariables() const { return Accessor<Variable>(_variables); }
+    const Accessor<PlanningProblem> getPlanningProblems() const { return Accessor<PlanningProblem>(_planningProblems); }
+
+    PlanRepository(const PlanRepository&) = delete;
+    PlanRepository(PlanRepository&&) = delete;
+
+    PlanRepository& operator=(const PlanRepository&) = delete;
+    PlanRepository& operator=(PlanRepository&&) = delete;
```
Comment (gauthamm) [inline alica_engine/include/engine/PlanRepository.h]: Doesn't seem like explicit delete of move operations is needed https://stackoverflow.com/questions/33096653/make-a-class-non-copyable-and-non-movable

File: alica_engine/include/engine/RunningPlan.h [hunk 4]
```diff
@@ -115,17 +116,14 @@ class RunningPlan : public enable_shared_from_this<RunningPlan> {
...
-                    robots);
+    void limitToRobots(const AgentSet& robots);
     std::shared_ptr<CycleManager> getCycleManagement();
     void revokeAllConstraints();
     void attachPlanConstraints();
-    bool recursiveUpdateAssignment(list<shared_ptr<SimplePlanTree>> spts,
-            std::vector<const supplementary::AgentID*>& availableAgents, list<const supplementary::AgentID*> noUpdates,
-            AlicaTime now);
+    bool recursiveUpdateAssignment(list<shared_ptr<SimplePlanTree>> spts, AgentSet& availableAgents,
+            list<const supplementary::AgentID*> noUpdates, AlicaTime now);
```
Comment (gauthamm) [inline alica_engine/include/engine/RunningPlan.h]: AgentList might be better

File: alica_engine/include/engine/RunningPlan.h [hunk 5]
```diff
@@ -115,17 +116,14 @@ class RunningPlan : public enable_shared_from_this<RunningPlan> {
     bool allChildrenStatus(PlanStatus ps);
     bool anyChildrenTaskSuccess();
     void activate();
-    EntryPoint* getActiveEntryPoint() const { return _activeEntryPoint; }
+    const EntryPoint* getActiveEntryPoint() const { return _activeEntryPoint; }
     void setActiveEntryPoint(EntryPoint* activeEntryPoint);
-    void limitToRobots(std::unordered_set<const supplementary::AgentID*, supplementary::AgentIDHash,
-            supplementary::AgentIDEqualsComparator>
-                    robots);
+    void limitToRobots(const AgentSet& robots);
```
Comment (gauthamm) [inline alica_engine/include/engine/RunningPlan.h]: AgentIDEqualsComparator can be removed from supplementary since its not used anywhere except for AgentIDManager. AgentIDmanager can also be modified to use AgentSet

File: alica_engine/src/engine/RunningPlan.cpp [hunk 6]
```diff
@@ -566,28 +567,29 @@ void RunningPlan::activate() {
...
+    // TODO: move this code into assignment to avoid the call to getAllRobots.
+    AgentSet curRobots;
+    _assignment->getAllRobots(curRobots);
+    for (const supplementary::AgentID* r : curRobots) {
+        if (find_if(robots.begin(), robots.end(), [&r](const supplementary::AgentID* id) { return *r == *id; }) ==
+                curRobots.end()) {
+            if (_activeState != nullptr && _assignment->getRobotStateMapping()->getStateOfRobot(r) == _activeState) {
                 recurse = true;
             }
             _assignment->removeRobot(r);
```
Comment (gauthamm) [inline alica_engine/src/engine/RunningPlan.cpp]: This internally does an erase on a vector which is ineffecient. Might be better for this to be a forward_list.

File: alica_engine/include/engine/BehaviourPool.h [hunk 7]
```diff
@@ -25,15 +25,15 @@ class BehaviourPool {
...
 
 private:
     /**
      * Manages behaviours used by the running ALICA program.
      * The key of the map is the behaviour configuration, which is created through the plan designer.
      * The value is the basic behaviour, which is the implementation of that behaviour.
      */
-    std::map<BehaviourConfiguration*, std::shared_ptr<BasicBehaviour>>* availableBehaviours;
+    std::map<const BehaviourConfiguration*, std::shared_ptr<BasicBehaviour>> _availableBehaviours;
     AlicaEngine* ae;
```
[the remark this answers is not in the document]
Comment (hendrik-skubch) [reply alica_engine/include/engine/BehaviourPool.h]: Yes, but not now. While at it, copy constructors should be deleted

File: alica_engine/include/engine/IAssignment.h [hunk 8]
```diff
@@ -24,8 +24,8 @@ class AssignmentCollection;
 class IAssignment {
 public:
     virtual ~IAssignment() {}
-    virtual shared_ptr<vector<const supplementary::AgentID*>> getRobotsWorking(EntryPoint* ep) = 0;
-    virtual shared_ptr<vector<const supplementary::AgentID*>> getRobotsWorking(long epid) = 0;
+    virtual const std::vector<const supplementary::AgentID*>* getRobotsWorking(const EntryPoint* ep) const = 0;
```
[the remark this answers is not in the document]
Comment (hendrik-skubch) [reply alica_engine/include/engine/IAssignment.h]: Yes, I missed some instances

File: alica_engine/include/engine/IAssignment.h [hunk 9]
```diff
@@ -35,36 +35,36 @@ class IAssignment {
...
-    virtual short getEntryPointCount() = 0;
+    virtual short getEntryPointCount() const = 0;
     /**
      * Returns all robot Ids working on the Task defined by ep and those which successfully completed it.
      * Behaviour is undefined if ep is not relevant or null.
      * @param ep The EntryPoint queried
      * @return A shared_ptr of a list of int of robot ids
      */
-    virtual shared_ptr<list<const supplementary::AgentID*>> getRobotsWorkingAndFinished(EntryPoint* ep) = 0;
+    virtual shared_ptr<list<const supplementary::AgentID*>> getRobotsWorkingAndFinished(const EntryPoint* ep) = 0;
```
[the remark this answers is not in the document]
Comment (hendrik-skubch) [reply alica_engine/include/engine/IAssignment.h]: Or better yet, we get rid of the lists. Lists are quite a bad datatype these days.

File: alica_engine/include/engine/PlanRepository.h (hunk 1 above)
Comment (hendrik-skubch) [reply alica_engine/include/engine/PlanRepository.h]: That is not universally true. Putting default(=empty) constructors (and destructors) in the cpp file can actually speed things up.

File: alica_engine/include/engine/SimplePlanTree.h (hunk 2 above)
Comment (hendrik-skubch) [reply alica_engine/include/engine/SimplePlanTree.h]: yes, that too

File: alica_engine/include/engine/PlanRepository.h (hunk 3 above)
Comment (hendrik-skubch) [reply alica_engine/include/engine/PlanRepository.h]: Maybe. But I do not trust compilers to be this correct wrt to the standard.
I know the microsoft compiler doesn't (or didn't) quite uphold that list.

File: alica_engine/include/engine/RunningPlan.h (hunk 4 above)
Comment (hendrik-skubch) [reply alica_engine/include/engine/RunningPlan.h]: No AgentList, lists are bad, we'll replace the datatypes one by one.

File: alica_engine/include/engine/RunningPlan.h (hunk 5 above)
Comment (hendrik-skubch) [reply alica_engine/include/engine/RunningPlan.h]: No, AgentIDEqualsComparator belongs with AgentID. And that package should not depend on alica, it would be circular.

File: alica_engine/src/engine/RunningPlan.cpp (hunk 6 above)
Comment (hendrik-skubch) [reply alica_engine/src/engine/RunningPlan.cpp]: Datatypes need to be reworked, yes.
But lists are the last option normally

File: alica_engine/include/engine/Types.h [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+using RoleVector = std::vector<const Role*>;
+using StateSet = std::vector<const State*>;
+using SuccessStateSet = std::vector<const SuccessState*>;
+using SyncTransitionSet = std::vector<const SyncTransition*>;
+using TaskSet = std::vector<const Task*>;
+using TransitionSet = std::vector<const Transition*>;
+using VariableSet = std::vector<const Variable*>;
+
+using BehaviourParameterMap = std::unordered_map<std::string, std::string>;
+}  // namespace alica
```
[the remark this answers is not in the document]
Comment (hendrik-skubch) [reply alica_engine/include/engine/Types.h]: done

File: alica_engine/include/engine/USummand.h [hunk 11]
```diff
@@ -45,29 +44,24 @@ class USummand {
         // init relevant entrypoint vector
         this->relevantEntryPoints.resize(this->relevantEntryPointIds.size());
         // find the right entrypoint for each id in relevant entrypoint id
-        map<long, EntryPoint*> elements = ae->getPlanRepository()->getEntryPoints();
-        EntryPoint* curEp;
         for (int i = 0; i < this->relevantEntryPoints.size(); ++i) {
```
[the remark this answers is not in the document]
Comment (hendrik-skubch) [reply alica_engine/include/engine/USummand.h]: yes, but not now

File: alica_engine/include/engine/USummand.h [hunk 12]
```diff
@@ -86,15 +80,13 @@ class USummand {
...
-    vector<long> relevantEntryPointIds;
-    /**
-     * Weight of this UtilitySummand
-     */
     double weight;
-    string name;
-    long id;
-    string info;
-    vector<EntryPoint*> relevantEntryPoints;
+    int64_t id;
```
[the remark this answers is not in the document]
Comment (hendrik-skubch) [reply alica_engine/include/engine/USummand.h]: no, it is an id to refer to an entrypoint... in future this should be made more explicit

File: alica_engine/include/engine/USummand.h [hunk 13]
```diff
@@ -86,15 +80,13 @@ class USummand {
...
+    std::vector<int64_t> relevantEntryPointIds;
```
[the remark this answers is not in the document]
Comment (hendrik-skubch) [reply alica_engine/include/engine/USummand.h]: yes, but not now

File: alica_engine/src/engine/model/TaskRepository.cpp [hunk 14]
```diff
@@ -9,38 +9,28 @@
...
+    _defaultTask = defaultTask;
 }
 
-string TaskRepository::getFileName() {
-    if (this->fileName.empty()) {
-        static string result = name + ".rdefset";
-        return result;
+std::string TaskRepository::getFileName() const {
+    if (_fileName.empty()) {
+        return getName() + ".rdefset";
```
[the remark this answers is not in the document]
Comment (hendrik-skubch) [reply alica_engine/src/engine/model/TaskRepository.cpp]: then you return .rdefset

File: alica_engine/include/engine/parser/PlanWriter.h [hunk 15]
```diff
@@ -44,51 +38,55 @@ class PlanWriter {
...
-    void saveRoleSet(RoleSet* r, string directory, string name);
-    tinyxml2::XMLDocument* createTaskRepositoryXMLDocument(TaskRepository* tr);
-    void saveTaskRepository(TaskRepository* tr, string name);
-    void saveTaskRepository(TaskRepository* tr, string directory, string name);
-    string getTempPlanDir();
-    void setTempPlanDir(string directory);
-    string getConfigPath();
-    vector<AlicaElement*>& getPlansToSave();
+    void saveSinglePlan(const Plan* p);
+    void saveSinglePlan(std::string directory, const Plan* p);
```
[the remark this answers is not in the document]
Comment (hendrik-skubch) [reply alica_engine/include/engine/parser/PlanWriter.h]: yes, but not now

File: alica_engine/src/engine/parser/PlanWriter.cpp [hunk 16]
```diff
@@ -522,7 +519,15 @@ string PlanWriter::getRelativeFileName(string file) {
         if (file.substr(file.size() - 4, 4) == ".beh" || file.substr(file.size() - 4, 4) == ".pty" ||
                 file.substr(file.size() - 4, 4) == ".pml" || file.substr(file.size() - 3, 3) == ".pp") {
             supplementary::SystemConfig* sc = supplementary::SystemConfig::getInstance();
-            string tfile = (*sc)["Alica"]->get<string>("Alica.PlanDir");
+            string tfile = (*sc)["Alica"]->get<string>("Alica.PlanDir", NULL);
+            supplementary::FileSystem::combinePaths(tfile, file);
```
Comment (gauthamm) [inline alica_engine/src/engine/parser/PlanWriter.cpp]: This line does nothing. The combinedPath is the returned string.

File: alica_engine/src/engine/parser/PlanWriter.cpp [hunk 17]
```diff
@@ -522,7 +519,15 @@ string PlanWriter::getRelativeFileName(string file) {
...
+            if (!supplementary::FileSystem::isPathRooted(tfile)) {
+                tfile = supplementary::FileSystem::combinePaths(this->configPath, tfile);
+            }
+            ufile = tfile;
+        } else if (file.substr(file.size() - 4, 4) == ".tsk") {
+            supplementary::SystemConfig* sc = supplementary::SystemConfig::getInstance();
+            string tfile = (*sc)["Alica"]->get<string>("Alica.MiscDir", NULL);
             supplementary::FileSystem::combinePaths(tfile, file);
```
Comment (gauthamm) [inline alica_engine/src/engine/parser/PlanWriter.cpp]: The returned string is not used

File: alica_engine/include/engine/constraintmodul/ConditionStore.h [hunk 18]
```diff
@@ -21,12 +20,12 @@ class ConditionStore {
...
-    void addCondition(Condition* con);
-    void removeCondition(Condition* con);
+    void addCondition(const Condition* con);
+    void removeCondition(const Condition* con);
 
     void acceptQuery(Query& query, std::shared_ptr<RunningPlan> rp) const;
-    std::list<Condition*> activeConditions;
-    std::map<Variable*, std::shared_ptr<std::vector<Condition*>>> activeVar2CondMap;
+    std::list<const Condition*> activeConditions;
+    std::map<const Variable*, std::shared_ptr<std::vector<const Condition*>>> activeVar2CondMap;
```
[the remark this answers is not in the document]
Comment (hendrik-skubch) [reply alica_engine/include/engine/constraintmodul/ConditionStore.h]: ConditionStore needs a bigger overhaul

File: alica_engine/include/engine/constraintmodul/IVariableSyncModule.h [hunk 19]
```diff
@@ -29,7 +29,7 @@ class IVariableSyncModule : public enable_shared_from_this<IVariableSyncModule>
 
     virtual void postResult(long vid, shared_ptr<vector<uint8_t>>& result) = 0;
     virtual shared_ptr<vector<shared_ptr<vector<shared_ptr<vector<uint8_t>>>>>> getSeeds(
-            shared_ptr<vector<Variable*>> query, shared_ptr<vector<shared_ptr<vector<double>>>> limits) = 0;
+            shared_ptr<vector<const Variable*>> query, shared_ptr<vector<shared_ptr<vector<double>>>> limits) = 0;
```
[the remark this answers is not in the document]
Comment (hendrik-skubch) [reply alica_engine/include/engine/constraintmodul/IVariableSyncModule.h]: VariableSyncModule is subject of my next little project. 
And no, let's not make it easier to use vector and shared ptr without any meaningful api together

File: alica_engine/src/engine/parser/PlanWriter.cpp (hunk 17 above)
Comment (gauthamm) [reply alica_engine/src/engine/parser/PlanWriter.cpp]: This function in general has unneeded variables and requires a cleanup

File: alica_engine/include/engine/model/Task.h [hunk 20]
```diff
@@ -8,37 +8,35 @@
...
-    const TaskRepository* taskRepository;
+    const TaskRepository* getTaskRepository() const {return _taskRepository;}
+    std::string toString() const;
 
-protected:
-    bool defaultTask;
+    static constexpr int64_t IDLEID = -1;  // For Task Id of an Idle EntryPoint...
+private:
+    void setTaskRepository(const TaskRepository* taskRepository);
+    friend ModelFactory;
```
[the remark this answers is not in the document]
Comment (hendrik-skubch) [reply alica_engine/include/engine/model/Task.h]: done

File: alica_engine/src/engine/planselector/PartialAssignment.cpp [hunk 21]
```diff
@@ -10,6 +10,9 @@
 #include "engine/planselector/DynCardinality.h"
 
 namespace alica {
+namespace {
+constexpr int INFINIT = std::numeric_limits<int>::max();
```
Comment (gauthamm) [inline alica_engine/src/engine/planselector/PartialAssignment.cpp]: INFINITY or INFINITE

File: alica_tests/src/test/etc/Alica.conf [hunk 22]
```diff
@@ -20,6 +20,7 @@
 	#relative to ES_ROOT env variable or absolute:
 	PlanDir = plans
 	RoleDir = roles
+    MiscDir = Misc
```
Comment (gauthamm) [inline alica_tests/src/test/etc/Alica.conf]: formating

File: alica_engine/src/engine/parser/PlanWriter.cpp (hunk 17 above)
Comment (hendrik-skubch) [reply alica_engine/src/engine/parser/PlanWriter.cpp]: yes it needs to have a cleanup. Not now though

File: alica_engine/src/engine/parser/PlanWriter.cpp (hunk 16 above)
Comment (hendrik-skubch) [reply alica_engine/src/engine/parser/PlanWriter.cpp]: good spot, let's fix this in the planwriter overhaul

File: alica_engine/include/engine/teammanager/Agent.h [hunk 23]
```diff
@@ -28,36 +28,41 @@ class Agent {
...
-    void setTimeLastMsgReceived(AlicaTime timeLastMsgReceived);
-    void setSuccess(AbstractPlan* plan, EntryPoint* entryPoint);
+    void setIgnored(const bool ignored) { _ignored = ignored; }
+    void setTimeLastMsgReceived(AlicaTime timeLastMsgReceived) { _timeLastMsgReceived = timeLastMsgReceived; }
+    void setSuccess(const AbstractPlan* plan, const EntryPoint* entryPoint);
     void setSuccessMarks(std::shared_ptr<SuccessMarks> successMarks);
-    Variable* getDomainVariable(std::string sort);
-    std::shared_ptr<std::list<EntryPoint*>> getSucceededEntryPoints(AbstractPlan* plan) const;
+    const Variable* getDomainVariable(const std::string& sort) const;
+    std::shared_ptr<std::list<const EntryPoint*>> getSucceededEntryPoints(const AbstractPlan* plan) const;
```
[the remark this answers is not in the document]
Comment (hendrik-skubch) [reply alica_engine/include/engine/teammanager/Agent.h]: no lists

File: alica_engine/src/engine/planselector/PartialAssignment.cpp (hunk 21 above)
Comment (hendrik-skubch) [reply alica_engine/src/engine/planselector/PartialAssignment.cpp]: done

File: alica_tests/src/test/etc/Alica.conf (hunk 22 above)
Comment (hendrik-skubch) [reply alica_tests/src/test/etc/Alica.conf]: done

File: alica_engine/include/engine/collections/RobotProperties.h [hunk 24]
```diff
@@ -38,9 +38,8 @@ class RobotProperties {
 protected:
     const supplementary::AgentID* agentId;
 
-    string defaultRole;
-    map<string, Characteristic*> characteristics;
-    map<long, Capability*> capabilities;
+    std::string defaultRole;
+    std::map<std::string, const Characteristic*> characteristics;
```
Comment (abolinsky) [inline alica_engine/include/engine/collections/RobotProperties.h]: characteristics are only read once from the configuration file, so ideally this should be const. The corresponding references to characteristics would need to be changed in **RobotProperties::readFromConfig**.

Unless this is forward thinking in the event robots are able to acquire new capabilities dynamically if we want to enable ALICA for use in real Transformers...

File: alica_engine/include/engine/model/Plan.h [hunk 25]
```diff
@@ -25,51 +24,57 @@ class PostCondition;
 class State;
 class SyncTransition;
 class Transition;
-
+class ModelFactory;
+class ExpressionHandler;
 /**
  * An ALICA plan
  */
 class Plan : public AbstractPlan {
```
Comment (abolinsky) [inline alica_engine/include/engine/model/Plan.h]: Add final specifier

Comment (abolinsky) [reply alica_engine/include/engine/model/Plan.h]: Also add override specifiers to any overridden methods below.

File: alica_engine/include/engine/model/PlanType.h [hunk 26]
```diff
@@ -13,28 +13,30 @@
...
 #include "AbstractPlan.h"
+#include "engine/Types.h"
 
-using namespace std;
 namespace alica {
 class Plan;
 class Parametrisation;
+class ModelFactory;
 
 class PlanType : public AbstractPlan {
```
Comment (abolinsky) [inline alica_engine/include/engine/model/PlanType.h]: Add final specifier and corresponding override specifiers for methods.

File: alica_engine/include/engine/model/PlanningProblem.h [hunk 27]
```diff
@@ -30,40 +28,44 @@ class PlanningProblem : public AbstractPlan {
 public:
```
Comment (abolinsky) [inline alica_engine/include/engine/model/PlanningProblem.h]: Add final specifier and corresponding override specifiers on methods.

File: alica_engine/include/engine/model/PostCondition.h [hunk 28]
```diff
@@ -9,18 +9,16 @@
...
 
 #include <string>
-#include <sstream>
 
 #include "Condition.h"
 
-using namespace std;
 namespace alica {
 
 class PostCondition : public Condition {
```
Comment (abolinsky) [inline alica_engine/include/engine/model/PostCondition.h]: final and overrides

File: alica_engine/include/engine/model/PreCondition.h [hunk 29]
```diff
@@ -8,31 +8,28 @@
...
-using namespace std;
 namespace alica {
 
-class RunningPlan;
+class ModelFactory;
 
 /**
  * A precondition guards a Plan or a Transition.
  */
 class PreCondition : public Condition {
```
Comment (abolinsky) [inline alica_engine/include/engine/model/PreCondition.h]: final and overrides

File: alica_engine/include/engine/model/Role.h [hunk 30]
```diff
@@ -8,38 +8,39 @@
...
 #include "AlicaElement.h"
 
-using namespace std;
 namespace alica {
 class RoleTaskMapping;
 class RoleDefinitionSet;
 class Characteristic;
+class ModelFactory;
 
 class Role : public AlicaElement {
```
Comment (abolinsky) [inline alica_engine/include/engine/model/Role.h]: final and overrides

File: alica_engine/include/engine/model/RoleDefinitionSet.h [hunk 31]
```diff
@@ -8,29 +8,28 @@
...
 
 #include "AlicaElement.h"
+#include "engine/Types.h"
 
-using namespace std;
 namespace alica {
-class Role;
+class ModelFactory;
 
 class RoleDefinitionSet : public AlicaElement {
```
Comment (abolinsky) [inline alica_engine/include/engine/model/RoleDefinitionSet.h]: final and overrides

File: alica_engine/include/engine/model/RoleSet.h [hunk 32]
```diff
@@ -8,35 +8,35 @@
...
-    list<RoleTaskMapping*>& getRoleTaskMappings();
-    void setRoleTaskMappings(const list<RoleTaskMapping*> roleTaskMappings);
-    long getUsableWithPlanId() const;
-    void setUsableWithPlanId(long usableWithPlanId);
+    std::string toString() const;
+    bool isDefault() const { return _isDefault; }
+    const std::vector<RoleTaskMapping*>& getRoleTaskMappings() const { return _roleTaskMappings; }
+    int64_t getUsableWithPlanId() const { return _usableWithPlanID; }
 
 protected:
```
Comment (abolinsky) [inline alica_engine/include/engine/model/RoleSet.h]: private

File: alica_engine/include/engine/model/RoleSet.h [hunk 33]
```diff
@@ -8,35 +8,35 @@
...
-#include <sstream>
+#include <vector>
 
 #include "AlicaElement.h"
 
-using namespace std;
 namespace alica {
 class RoleTaskMapping;
 
 class RoleSet : public AlicaElement {
```
Comment (abolinsky) [inline alica_engine/include/engine/model/RoleSet.h]: final and overrides

File: alica_engine/include/engine/model/RoleTaskMapping.h [hunk 34]
```diff
@@ -8,32 +8,31 @@
...
 
 #include "AlicaElement.h"
 
-using namespace std;
 namespace alica {
 
 class Role;
+class ModelFactory;
 
 class RoleTaskMapping : public AlicaElement {
```
Comment (abolinsky) [inline alica_engine/include/engine/model/RoleTaskMapping.h]: private and overrides

File: alica_engine/include/engine/model/RuntimeCondition.h [hunk 35]
```diff
@@ -9,20 +9,18 @@
...
 #include <string>
-#include <sstream>
 
 #include "Condition.h"
 
-using namespace std;
 namespace alica {
 class RunningPlan;
 
 class RuntimeCondition : public Condition {
```
Comment (abolinsky) [inline alica_engine/include/engine/model/RuntimeCondition.h]: final and overrides

File: alica_engine/include/engine/model/State.h [hunk 36]
```diff
@@ -12,80 +12,84 @@
...
  */
 class State : public AlicaElement {
 public:
+    enum StateType { NORMAL, SUCCESS, FAILURE };
     State();
-    State(long id);
+    State(StateType t);
+    State(int64_t id);
     virtual ~State();
-    virtual string toString() const;
```
Comment (abolinsky) [inline alica_engine/include/engine/model/State.h]: no more toString?

File: alica_engine/include/engine/model/SyncTransition.h [hunk 37]
```diff
@@ -13,37 +13,44 @@
...
+#include "engine/Types.h"
 
-using namespace std;
 namespace alica {
 
 class Plan;
 class Transition;
+class ModelFactory;
 
 class SyncTransition : public AlicaElement {
```
Comment (abolinsky) [inline alica_engine/include/engine/model/SyncTransition.h]: final and overrides

File: alica_engine/include/engine/model/SuccessState.h [hunk 38]
```diff
@@ -22,7 +20,7 @@ class SuccessState : public TerminalState {
 public:
```
Comment (abolinsky) [inline alica_engine/include/engine/model/SuccessState.h]: final and overrides

File: alica_engine/include/engine/model/TerminalState.h [hunk 39]
```diff
@@ -12,19 +12,20 @@
 
 namespace alica {
 class PostCondition;
-
+class ModelFactory;
 /**
  * A terminal state within a plan. Indicates termination of the corresponding task
  */
 class TerminalState : public State {
```
Comment (abolinsky) [inline alica_engine/include/engine/model/TerminalState.h]: I guess final and overrides can be another PR entirely and for another time.

File: alica_engine/include/engine/parser/PlanWriter.h [hunk 40]
```diff
@@ -44,51 +38,55 @@ class PlanWriter {
 public:
     PlanWriter(AlicaEngine* ae, PlanRepository* rep);
     virtual ~PlanWriter();
```
Comment (abolinsky) [inline alica_engine/include/engine/parser/PlanWriter.h]: should not be virtual

File: alica_engine/include/engine/planselector/TaskAssignment.h [hunk 41]
```diff
@@ -31,31 +31,30 @@ class PartialAssignmentPool;
  * Represents an instance of an assignment problem for one plan or a plantype.
  * All parameters, which are static for this problem, are stored here.
  */
-class TaskAssignment : virtual public ITaskAssignment {
+class TaskAssignment final : public ITaskAssignment {
```
Comment (abolinsky) [inline alica_engine/include/engine/planselector/TaskAssignment.h]: do we even need an ITaskAssignment class to inherit from?

File: alica_engine/include/engine/teammanager/Agent.h [hunk 42]
```diff
@@ -28,36 +28,41 @@ class Agent {
     friend ::alica::TeamObserver;
```
[the remark this answers is not in the document]
Comment (abolinsky) [reply alica_engine/include/engine/teammanager/Agent.h]: @gauthamm the prefixed :: denotes the root namespace to avoid any possible namespace collisions. Kind of like giving an absolute path instead of a relative path. More safe and less prone to error in cases of there being another namespace of the same name, say, in another library.

File: alica_engine/include/engine/teammanager/Agent.h [hunk 43]
```diff
@@ -28,36 +28,41 @@ class Agent {
...
 
 public:
-    virtual ~Agent();
+    virtual ~Agent() {
```
Comment (abolinsky) [inline alica_engine/include/engine/teammanager/Agent.h]: no need for virtual destructor.

File: alica_engine/src/engine/Assignment.cpp [hunk 44]
```diff
@@ -53,33 +56,19 @@ Assignment::Assignment(PartialAssignment* pa) {
...
+        epRobotsMapping->setEp(i, assCol->getEp(i));
         // copy robots
-        auto robots = assCol->getRobots(i);
-        curRobots = make_shared<vector<const supplementary::AgentID*>>();
-        for (auto& rob : *robots) {
-            curRobots->push_back(rob);
-        }
-
+        *(epRobotsMapping->editRobots(i)) = *(assCol->getRobots(i));
         // set the robots
```
Comment (abolinsky) [inline alica_engine/src/engine/Assignment.cpp]: I assume setRobots is called in editRobots?

File: alica_engine/src/engine/Assignment.cpp [hunk 45]
```diff
@@ -88,87 +77,77 @@ Assignment::Assignment(Plan* p, shared_ptr<AllocationAuthorityInfo> aai) {
...
 
-        curRobots = make_shared<vector<const supplementary::AgentID*>>();
+        // curRobots = make_shared<vector<const supplementary::AgentID*>>();
         for (auto epRobots : aai->entryPointRobots) {
             // find the right entrypoint
-            if (epRobots.entrypoint == epPair.second->getId()) {
+            if (epRobots.entrypoint == ep->getId()) {
                 // copy robots
+                AgentSet* curRobots = this->epRobotsMapping->editRobots(i);
+                curRobots->clear();
```
Comment (abolinsky) [inline alica_engine/src/engine/Assignment.cpp]: what's going on here? Clearing after populating?

Comment (hendrik-skubch) [reply alica_engine/src/engine/Assignment.cpp]: good catch! done

File: alica_engine/src/engine/Assignment.cpp (hunk 44 above)
Comment (hendrik-skubch) [reply alica_engine/src/engine/Assignment.cpp]: removed superfluous comment

File: alica_engine/include/engine/planselector/TaskAssignment.h (hunk 41 above)
Comment (hendrik-skubch) [reply alica_engine/include/engine/planselector/TaskAssignment.h]: Currently it bridges between TaskAssignment and PartialAssignment. We may be able to lift that though.

File: alica_engine/include/engine/teammanager/Agent.h (hunk 43 above)
Comment (hendrik-skubch) [reply alica_engine/include/engine/teammanager/Agent.h]: done

File: alica_engine/include/engine/parser/PlanWriter.h (hunk 40 above)
Comment (hendrik-skubch) [reply alica_engine/include/engine/parser/PlanWriter.h]: done

File: alica_engine/include/engine/model/TerminalState.h (hunk 39 above)
Comment (hendrik-skubch) [reply alica_engine/include/engine/model/TerminalState.h]: Added override, but skip on final ;)

File: alica_engine/include/engine/model/State.h (hunk 36 above)
Comment (hendrik-skubch) [reply alica_engine/include/engine/model/State.h]: it was meaningless

File: alica_engine/include/engine/model/RoleTaskMapping.h (hunk 34 above)
Comment (hendrik-skubch) [reply alica_engine/include/engine/model/RoleTaskMapping.h]: it's private...

File: alica_engine/include/engine/model/RoleSet.h (hunk 32 above)
Comment (hendrik-skubch) [reply alica_engine/include/engine/model/RoleSet.h]: done

File: alica_engine/include/engine/collections/RobotProperties.h (hunk 24 above)
Comment (hendrik-skubch) [reply alica_engine/include/engine/collections/RobotProperties.h]: maybe next time....

File: alica_engine/src/engine/collections/AssignmentCollection.cpp [hunk 46]
```diff
@@ -51,81 +87,111 @@ shared_ptr<vector<const supplementary::AgentID*>> AssignmentCollection::getRobot
...
-    }
+void AssignmentCollection::setEp(short index, const EntryPoint* ep) {
+    assert(index < _numEps);
+    _entryPoints[index] = ep;
 }
 
+void AssignmentCollection::addRobot(const supplementary::AgentID* id, const EntryPoint* e) {
+    for (int i = 0; i < _numEps; ++i) {
+        if (_entryPoints[i] == e) {
+            _robotIds[i].push_back(id);
```
Comment (abolinsky) [inline alica_engine/src/engine/collections/AssignmentCollection.cpp]: the way addRobot and removeRobot are implemented tells me that _robotIds should be a map instead of a vector.

File: alica_engine/src/engine/model/Quantifier.cpp [hunk 47]
```diff
@@ -10,105 +10,37 @@
...
-
-    if (scopeIsPlan) {
-        this->plan = (Plan*) element;
-    } else if (scopeIsEntryPoint) {
-        this->entryPoint = (EntryPoint*) element;
-    } else if (scopeIsState) {
-        this->state = (State*) element;
+void Quantifier::setScope(const AlicaElement* element) {
+    _scope = element;
+    if (dynamic_cast<const EntryPoint*>(element) != nullptr) {
```
Comment (abolinsky) [inline alica_engine/src/engine/model/Quantifier.cpp]: these dynamic casts shouldn't be here. These are only nullptr checks.
